### PR TITLE
Improve Classic NAS mutator

### PR DESCRIPTION
**Pull request by ultmaster**
1. Use mutator instead of BaseMutator as base class to simplify the code.
2. Add more validation to detect possible errors and give reasonable error messages.
3. Fix bug when search space path is an absolute path.

New utility functions in preparation for SPOS:

1. ModelCheckpoint
2. Add access shortcut for average meters.

### CODE REVIEW

File: src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py [hunk 1]
```diff
@@ -37,70 +44,102 @@ def __init__(self, model):
...
+        model : PyTorch model
             user's model with search space (e.g., LayerChoice, InputChoice) embedded in it
         """
         super(ClassicMutator, self).__init__(model)
-        self.chosen_arch = {}
-        self.search_space = self._generate_search_space()
-        if 'NNI_GEN_SEARCH_SPACE' in os.environ:
+        self._chosen_arch = {}
+        self._search_space = self._generate_search_space()
+        if "NNI_GEN_SEARCH_SPACE" in os.environ:
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py]: change the "NNI_GEN_SEARCH_SPACE" as an enum string?

File: src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py [hunk 2]
```diff
@@ -37,70 +44,102 @@ def __init__(self, model):
...
-        for key, val in self.search_space.items():
-            if val['_type'] == 'layer_choice':
-                choices = val['_value']
-                chosen_arch[key] = {'_value': choices[0], '_idx': 0}
-            elif val['_type'] == 'input_choice':
-                choices = val['_value']['candidates']
-                n_chosen = val['_value']['n_chosen']
-                chosen_arch[key] = {'_value': choices[:n_chosen], '_idx': list(range(n_chosen))}
+        for key, val in self._search_space.items():
+            if val["_type"] == "layer_choice":
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py]: same here.

File: src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py (hunk 1 above)
Comment (ultmaster) [reply src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py]: updated

File: src/sdk/pynni/nni/nas/pytorch/callbacks.py [hunk 3]
```diff
@@ -51,4 +54,23 @@ def __init__(self, checkpoint_dir, every="epoch"):
...
-        self.trainer.export(os.path.join(self.checkpoint_dir, "epoch_{}.json".format(epoch)))
+        dest_path = os.path.join(self.checkpoint_dir, "epoch_{}.json".format(epoch))
+        _logger.info("Saving architecture to %s", dest_path)
+        self.trainer.export(dest_path)
+
+
+class ModelCheckpoint(Callback):
+    def __init__(self, checkpoint_dir, every="epoch"):
+        super().__init__()
+        assert every == "epoch"
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/callbacks.py]: support only one specific value? so why we make it a parameter?

Comment (ultmaster) [reply src/sdk/pynni/nni/nas/pytorch/callbacks.py]: The roadmap is to support checkpoint on every mini-batch. But it's not implemented yet.

Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/callbacks.py]: please remove, because it is confusing.

File: src/sdk/pynni/nni/nas/pytorch/callbacks.py [hunk 4]
```diff
@@ -51,4 +54,23 @@ def __init__(self, checkpoint_dir, every="epoch"):
         os.makedirs(self.checkpoint_dir, exist_ok=True)
 
     def on_epoch_end(self, epoch):
-        self.trainer.export(os.path.join(self.checkpoint_dir, "epoch_{}.json".format(epoch)))
+        dest_path = os.path.join(self.checkpoint_dir, "epoch_{}.json".format(epoch))
+        _logger.info("Saving architecture to %s", dest_path)
+        self.trainer.export(dest_path)
+
+
+class ModelCheckpoint(Callback):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/callbacks.py]: can checkpointed model be used for resume? e.g., learning rate, epoch number

File: src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py [hunk 5]
```diff
@@ -37,70 +48,102 @@ def __init__(self, model):
...
 
-    def _validate_chosen_arch(self):
-        pass
+        if trial_env_vars.NNI_PLATFORM is None:
+            logger.warning("This is in standalone mode, the chosen are the first one(s).")
+            self._chosen_arch = self._standalone_generate_chosen()
+        else:
+            # get chosen arch from tuner
+            self._chosen_arch = nni.get_next_parameter()
+        self._cache = self.sample_final()
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py]: what is `_cache` used for in classic_nas?

File: src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py [hunk 6]
```diff
@@ -37,70 +48,102 @@ def __init__(self, model):
...
+            value = data["_value"]
+            idx = data["_idx"]
+            search_space_ref = self._search_space[mutable.key]["_value"]
+            if isinstance(mutable, LayerChoice):
+                # doesn't support multihot for layer choice yet
+                onehot_list = [False] * mutable.length
+                assert 0 <= idx < mutable.length and search_space_ref[idx] == value, \
+                    "Index '{}' in search space '{}' is not '{}'".format(idx, search_space_ref, value)
+                onehot_list[idx] = True
+                result[mutable.key] = torch.tensor(onehot_list, dtype=torch.bool)  # pylint: disable=not-callable
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py]: suggest to extract this part into a private member function

File: src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py [hunk 7]
```diff
@@ -37,70 +48,102 @@ def __init__(self, model):
...
+            elif isinstance(mutable, InputChoice):
+                multihot_list = [False] * mutable.n_candidates
+                for i, v in zip(idx, value):
+                    assert 0 <= i < mutable.n_candidates and search_space_ref[i] == v, \
+                        "Index '{}' in search space '{}' is not '{}'".format(i, search_space_ref, v)
+                    assert not multihot_list[i], "'{}' is selected twice in '{}', which is not allowed.".format(i, idx)
+                    multihot_list[i] = True
+                result[mutable.key] = torch.tensor(multihot_list, dtype=torch.bool)  # pylint: disable=not-callable
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py]: suggest to extract this part into a private member function

File: src/sdk/pynni/nni/nas/pytorch/callbacks.py (hunk 4 above)
Comment (ultmaster) [reply src/sdk/pynni/nni/nas/pytorch/callbacks.py]: This part is in another PR. And no, not supported.

File: src/sdk/pynni/nni/nas/pytorch/callbacks.py (hunk 3 above)
Comment (ultmaster) [reply src/sdk/pynni/nni/nas/pytorch/callbacks.py]: fixed in another PR.

File: src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py (hunk 5 above)
Comment (ultmaster) [reply src/sdk/pynni/nni/nas/pytorch/classic_nas/mutator.py]: `_cache` is stored so that `on_forward_layer_choice` implemented in Mutator can retrieve the decision.
